File: 5. FactorialIterativeRecursive.cpp
/*
Task: Implementation and Time Analysis of Factorial Program using Iterative and Recursive Method
*/
#include <iostream>
#include <ctime>
using namespace std;

long long factorialIterative(int n)
{
    long long res = 1;
    for (int i = 2; i <= n; ++i) res *= i;
    return res;
}
long long factorialRecursive(int n)
{
    if (n <= 1)
        return 1;
    return n * factorialRecursive(n - 1);
}
int main()
{
    int n = 20;
    clock_t start = clock();
    long long res = factorialIterative(n);
    clock_t end = clock();
    cout << "Iterative: " << res << ", Time: " << (double)(end - start) / CLOCKS_PER_SEC << "s\n";
    start = clock();
    res = factorialRecursive(n);
    end = clock();
    cout << "Recursive: " << res << ", Time: " << (double)(end - start) / CLOCKS_PER_SEC << "s\n";
    return 0;
}
/*
Time Complexity: O(n)
Space Complexity: Iterative O(1), Recursive O(n)
*/

File: 5. FactorialIterativeRecursive.py
"""
Task: Implementation and Time Analysis of Factorial Program using Iterative and Recursive Method
"""
import time
def factorial_iterative(n):
    res = 1
    for i in range(2, n+1):
        res *= i
    return res
def factorial_recursive(n):
    if n <= 1:
        return 1
    return n * factorial_recursive(n-1)
if __name__ == "__main__":
    n = 20
    start = time.time()
    res = factorial_iterative(n)
    end = time.time()
    print(f"Iterative: {res}, Time: {end-start:.6f}s")
    start = time.time()
    res = factorial_recursive(n)
    end = time.time()
    print(f"Recursive: {res}, Time: {end-start:.6f}s")
"""
Time Complexity: O(n)
Space Complexity: Iterative O(1), Recursive O(n)
"""
